File: exercises/generate_parentheses_22.java
import java.util.ArrayList;
import java.util.List;

//Original Solution Recursive
class Solution {
    List<String> ans = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
        rec("", n, n);
        return this.ans;
    }
    void rec (String s, int open, int close) {
        if (open == 0 && close == 0) {
            ans.add(s);
            return;
        }
        if (open > 0) {
            rec(s+"(", open-1, close);
        }
        if ( close < open) {
            rec(s+")", open, close-1);
        }
    }
}

//import java.util.Stack;
//class Solution {
//
//    public List<String> generateParenthesis(int n) {
//        List<String> ans = new ArrayList<>();
//        var stack = new Stack<StackElement>();
//        stack.add(new StackElement("", n, n));
//        while (!stack.empty()) {
//            var element = stack.pop();
//            String comb = element.comb;
//            int open = element.open;
//            int close = element.close;
//            //if (open == 0 && close != 0) return;
//            if (open == 0 && close == 0) {
//                ans.add(comb);
//                continue;
//            }
//            if (open > 0) {
//                stack.push(new StackElement(comb + "(", open - 1, close));
//            }
//            if (open < close) {
//                stack.push(new StackElement(comb + ")", open, close - 1));
//            }
//        }
//        return ans;
//    }
//}
//
class Element {
    String comb = "";
    int open = 0;
    int close = 0;

    public Element(String comb, int open, int close) {
        this.comb = comb;
        this.open = open;
        this.close = close;
    }
}

//class Solution {
//    public List<String> generateParenthesis(int n) {
//        List<String> ans = new ArrayList<>();
//        List<Element> queue = new ArrayList<Element>();
//        queue.add(new Element("", n, n));
//        int f = 0;  // f is first element
//        int l = 1;   // l is length
//        while (f < l) {
//            var element = queue.get(f);
//            String comb = element.comb;
//            int open = element.open;
//            int close = element.close;
//            if (open == 0 && close == 0) {
//                ans.add(comb);
//                f++;
//                continue;
//            }
//            if (open > 0) {
//                queue.add(new Element(comb + "(", open - 1, close));
//                l++;
//            }
//            if (open < close) {
//                queue.add(new Element(comb + ")", open, close - 1));
//                l++;
//            }
//            f++;
//        }
//
//        return ans;
//    }
//}

// Prompt
//Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
//
//
//
//Example 1:
//
//Input: n = 3
//Output: ["((()))","(()())","(())()","()(())","()()()"]
//
//Example 2:
//
//Input: n = 1
//Output: ["()"]
//
//
//
//Constraints:
//
//    1 <= n <= 8

File: exercises/generate_parentheses_22.py
from typing import List

# Original Solution using recursion ( recursion uses Stack under the hood )
# class Solution:
#     res: List[str] = []
#     def generateParenthesis(self, n: int) -> List[str]:
#         self.res = []
#         self.recurse(n, n, "")
#         return self.res
#
#     def recurse (self, opening: int, closing: int, comb: str):
#         if opening == 0 and closing == 0:
#             self.res.append(comb)
#             return
#         if opening > 0:
#             self.recurse(opening-1, closing, comb + "(")
#         if closing > opening:
#             self.recurse(opening, closing - 1, comb + ")")


# Solution using Stack ( Managing a stack myself )
# This is actually a very crude translation from recursive to iterative (but it works wonderfully)
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = [(n, n, "")]
        res = []

        while stack:
            opening, closing, comb = stack.pop()
            if opening == 0 and closing == 0:
                res.append(comb)
                continue
            if closing > opening:
                stack.append((opening, closing - 1, comb + ")"))

            if opening > 0:
                stack.append((opening - 1, closing, comb + "("))
        return res


# def generateParenthesis(self, n: int) -> List[str]:
#     ans = []
#     queue = [('(', 1, 0)]  #(str, open_nr, close_nr)
#     f, l = 0, 1
#     while f<l:
#         comb, op, cl = queue[f]
#         if op+cl==2*n:
#             ans.append(comb)
#         elif op < n:
#             queue.append((comb+'(', op+1, cl))
#             l+=1
#         elif op >= cl:
#             queue.append((comb+')', op, cl+1))
#             l+=1
#         f+=1
#     return ans

# class Solution:
#     def generateParenthesis(self, n: int) -> List[str]:
#         ans = []
#         queue = [('(', 1, 0)]  #(str, open_nr, close_nr)
#         f, l = 0, 1
#         while f < l:
#             comb, op, cl = queue[f]
#             if op+cl==2*n:
#                 ans.append(comb)
#                 f+=1
#                 continue
#             if op < n:
#                 queue.append((comb+'(', op+1, cl))
#                 l += 1
#             if op > cl:
#                 queue.append((comb+')', op, cl+1))
#                 l += 1
#             f+=1
#         return ans


# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
#
#
#
# Example 1:
#
# Input: n = 3
# Output: ["((()))","(()())","(())()","()(())","()()()"]
#
# Example 2:
#
# Input: n = 1
# Output: ["()"]
#
#
#
# Constraints:
#
#     1 <= n <= 8
